### DISCUSSION TITLE
How to get all types and each rules required fields

### DISCUSSION THREAD

**Discussion by Lemour-sudo** [eclipse-langium/langium#1380] [Q&A]
I have parser rules that expect multiple fields to be included: `Assert: 'assert' '(' value1=STRING possibleValues+=STRING ')';`.
For each parser rule I want to get its metadata that will tell me the type and the required fields. In this example, I would expect something like:

```
{
    name: 'Assert',
    parameters: {
        value1: string;
        value2: string[];
    }
}
```

I did come across the mandatory property returned as part of a type's metadata with the `getTypeMetaData` method in the generated "ast.ts" file. However, the mandatory property is just a list of one required field:
```
{
    name: 'Assert',
    mandatory: [
        { name: 'value2', type: 'array' }
    ]
}
```

First, `value1` is not included. Second, from the above metadata, it is not clear what type `value2` is an array of.

Is it possible in Langium to get all a parser rule's required fields like in the above example?

**Comment by msujew** [comment]
Hey @Lemour-sudo,

in the current [`next`](https://www.npmjs.com/package/langium/v/3.0.0-next.e78aeba) release (see also https://github.com/eclipse-langium/langium/pull/1165), all properties are now being exposed by the API. Note that the `getTypeMetaData` in v2 is mainly an internal API: It tells the parser which fields on an AST node can be filled with reasonable defaults. This only includes arrays (defaulting to `[]`) and booleans (defaulting to `false`). Filling other fields would lead to unexpected side effects.

Note that you can always get the fields of an AST node by using `Object.keys(node)` and filtering out the fields starting with `$`. This assumes that the node has been parsed correctly without parser errors.

**Reply by Lemour-sudo** [reply]
Thanks for the reply.

I see that AST is an interface in "ast.ts", how can I treat it as an object instead in order to go through its keys as you suggested?

**Reply by msujew** [reply]
Interfaces are a compile time construct of TypeScript. They are deleted when generating the JavaScript code, since JS doesn't have a notion of an interface. There's also no corresponding object that gets generated instead.

**Reply by Lemour-sudo** [reply]
I was actually referring to the `node` object you mentioned in the first reply. How can I find this node in Langium?

**Reply by msujew** [reply]
I meant an instance of an `AstNode` interface object (i.e. any of your AST nodes). Assuming there was no parser error, they should contain all non-nullable elements.
